### PR TITLE
feat: use-auth-client react hook

### PR DESCRIPTION
**Pull request by krpeacock**
# Description

Introduces a new library: `@dfinity/use-auth-client`

Simplifies managing authentication using best practices with hooks and effects.

`Logout` and `onIdle` also ends session for Actors managed by the auth hook.

Example app live at https://5ibdo-haaaa-aaaab-qajia-cai.icp0.io/

# How Has This Been Tested?

New unit test and an example project. More testing would probably be worthwhile

# Checklist:

- [X] My changes follow the guidelines in [CONTRIBUTING.md](https://github.com/dfinity/agent-js/blob/main/CONTRIBUTING.md).
- [X] The title of this PR complies with [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/).
- [X] I have edited the CHANGELOG accordingly.
- [X] I have made corresponding changes to the documentation.

### CODE REVIEW

File: packages/use-auth-client/lib/cjs/packages/use-auth-client/src/index.js
Comment (dfx-json) [inline packages/use-auth-client/lib/cjs/packages/use-auth-client/src/index.js:1]: Should generated files be checked in?

File: packages/use-auth-client/lib/esm/packages/use-auth-client/src/use-auth-client.js [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    function login() {
+        return new Promise((resolve, reject) => {
+            if (authClient) {
+                const callback = options?.loginOptions?.onSuccess;
+                const errorCb = options?.loginOptions?.onError;
+                authClient.login({
+                    ...options?.loginOptions,
+                    onSuccess: (successResponse) => {
+                        if (successResponse !== undefined) {
+                            callback?.(successResponse);
```
Comment (dfx-json) [inline packages/use-auth-client/lib/esm/packages/use-auth-client/src/use-auth-client.js:36]: Should the promise be resolved here too?

**Review by dfx-json — APPROVED** [review]

File: packages/use-auth-client/lib/cjs/packages/use-auth-client/src/index.js
Comment (krpeacock) [reply packages/use-auth-client/lib/cjs/packages/use-auth-client/src/index.js:1]: no, they should be removed

File: packages/use-auth-client/src/use-auth-client.ts [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+      if (authClient) {
+        const callback = options?.loginOptions?.onSuccess;
+        const errorCb = options?.loginOptions?.onError;
+        authClient.login({
+          ...options?.loginOptions,
+          onSuccess: (successResponse?: InternetIdentityAuthResponseSuccess) => {
+            setIsAuthenticated(true);
+            setIdentity(authClient.getIdentity());
+            if (successResponse !== undefined) {
+              callback?.(successResponse);
```
Comment (dfx-json) [inline packages/use-auth-client/src/use-auth-client.ts:115]: same question — why shouldn't the promise resolve here as well?

Comment (krpeacock) [reply packages/use-auth-client/src/use-auth-client.ts:115]: Yes, you're right

**Pull request merged**